### PR TITLE
Implement new Vue Virtual Grid

### PR DESCRIPTION
**Pull request by Mikescops**
This implement the new virtual grid lib I created: https://www.npmjs.com/package/vue-virtual-grid

Also fast forward #125

And finally fix a bug with title rendering and the new sidebar.

Please take time to QA this to make sure there is no regression I couldn't catch.

Preview of it:

image

and:

image

### CODE REVIEW

File: src/components/SeparatorVirtualGrid.vue [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+export default {
+	name: 'SeparatorVirtualGrid',
+
+	inheritAttrs: false,
+
+	props: {
+		item: {
+			type: Object,
+			required: true,
+			injected: {
```
Comment (skjnldsv) [inline src/components/SeparatorVirtualGrid.vue:18]: Does that exists? Never saw it before: `injected`
https://vuejs.org/v2/api/#props

File: src/components/FileVirtualGrid.vue [hunk 2]
```diff
@@ -0,0 +1,143 @@
+
+<template>
+	<a :class="{'file--clear': !loaded}"
```
Comment (skjnldsv) [inline src/components/FileVirtualGrid.vue:3]: Why do we need a new dedicated template?
Ideally we should keep using the same File component that any other Views. That way any change is visible and propagated to all! :thinking:

File: src/views/Albums.vue [hunk 3]
```diff
@@ -30,30 +30,31 @@
...
-				:key="dir.fileid"
-				v-bind="dir"
-				:show-shared="showShared" />
-			<File v-for="file in fileList"
-				:key="file.fileid"
-				:list="fileList"
-				v-bind="file" />
-		</template>
-	</Grid>
+		<Grid>
```
Comment (skjnldsv) [inline src/views/Albums.vue:39]: I feel like this is not related to this pr, no?
What is this trying to fix?

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:18]: it's a custom property from my lib

File: src/components/FileVirtualGrid.vue (hunk 2 above)
Comment (Mikescops) [reply src/components/FileVirtualGrid.vue:3]: Because props are passed from the lib and as I didn't want to change the other views Grid system i created a new one, but i can remove it for sure

File: src/views/Albums.vue (hunk 3 above)
Comment (Mikescops) [reply src/views/Albums.vue:39]: I changed rendering of navigation to fit the new Virtual Grid but it was impacting the other views so I had to make a quick fix there also.

If we consider to add the virtual grid everywhere then it's fine!

**Review by Mikescops — COMMENTED** [review]
I found few bugs:
- (not related to this PR) XML parsing has an issue when there is no images
- when resizing the window the computation goes wrong for full width items (this has to be fixed in the virtual grid lib)
- the viewer is locked to folders once again (maybe we will also have to edit the virtual grid lib to manually trigger the next batch of data)

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (skjnldsv) [reply src/components/SeparatorVirtualGrid.vue:18]: But it's not passed. you cannot nest props in Vue.
If you needed to pass a default, then you need to use the default key, `injected` means nothing to vue here. :)

Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:18]: ah right, i think i'm mixing things up with Typescript, you're right i only need to have the root property and i can still access to item.injected.xxxx

Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:18]: This is fixed

**Comment by Mikescops** [comment]
>     * when resizing the window the computation goes wrong for full width items (this has to be fixed in the virtual grid lib)
> 
>     * the viewer is locked to folders once again (maybe we will also have to edit the virtual grid lib to manually trigger the next batch of data)

Both issues are now fixed + the ones reported by @skjnldsv 

@jancborchardt if you can give a try

**Review by jancborchardt — CHANGES_REQUESTED** [review]
Very nice! Only some small feedback:
- With the new date headings, the "Your photos" heading is not necessary anymore and can be cut :) But make sure there’s some nice whitespace on top of the first date heading
- The date headings could do with some more whitespace above them, let’s say double of what’s there now (so 72px)
- I get this strange preview glitch when scrolling (in Firefox) where already loaded pictures fall back to displaying only the filetype icon:
Photos scroll loading glitch

**Comment by Mikescops** [comment]
I have the same glitches and I'm not sure why they are there, and it seems somehow that the old grid what already doing that but less visible.

I'll try to find why and fix the other spacing issues.

**Comment by Mikescops** [comment]
@jancborchardt it's fixed :smile: 
thanks @skjnldsv

**Comment by skjnldsv** [comment]
> it's fixed

Congrats! Was not that easy :p

File: src/components/FileVirtualGrid.vue [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			})
+		},
+
+		onLoad() {
+			this.loaded = true
+		},
+
+		filterOnlyViewableFiles(list) {
+			return list.filter(item => item.injected.filename).map(item => item.injected)
+		},
```
Comment (skjnldsv) [inline src/components/FileVirtualGrid.vue:91]: I don't understand the purpose of this function?
Why filtering the filenames? 
Couldnt this directly be managed by the loadMore function?

Comment (Mikescops) [reply src/components/FileVirtualGrid.vue:91]: for the loadmore it's not necessary, true, but for the current list it is because i'm injecting title/separators

Comment (Mikescops) [reply src/components/FileVirtualGrid.vue:91]: i was wrong it's necessary for both

File: src/views/Timeline.vue [hunk 5]
```diff
@@ -143,47 +123,69 @@ export default {
...
 			const { request, cancel } = cancelableRequest(getPhotos)
 			this.cancelRequest = cancel
 
 			try {
-				// get content and current folder info
 				const files = await request(this.onlyFavorites, {
-					page: this.page,
-					perPage: this.gridConfig.count * 5, // we load 5 rows,
+					page: params.offset,
+					perPage: 30,
```
Comment (skjnldsv) [inline src/views/Timeline.vue:143]: Use the grid config

File: src/views/Timeline.vue [hunk 6]
```diff
@@ -205,63 +211,29 @@ export default {
...
 		},
 
 	},
 
 }
 </script>
+
+<style lang="scss" scoped>
+.grid-container {
+	padding: 66px;
```
Comment (skjnldsv) [inline src/views/Timeline.vue:237]: Padding is variable.
Please check https://github.com/nextcloud/photos/blob/b501e72f31bbaf90d89e3c55ef2438b002244d0b/src/assets/grid-sizes.js

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (skjnldsv) [inline src/components/SeparatorVirtualGrid.vue:18]: Still here

File: src/components/SeparatorVirtualGrid.vue [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+				type: Number,
+				required: true,
+			},
+		},
+	},
+}
+</script>
+
+<style lang="scss" scoped>
+.grid-title {
```
Comment (skjnldsv) [inline src/components/SeparatorVirtualGrid.vue:40]: Grid title should be using the grid config sizing (padding, margins...)

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (skjnldsv) [inline src/components/SeparatorVirtualGrid.vue:18]: `injected` needs to go away.

File: src/views/Tags.vue [hunk 8]
```diff
@@ -27,37 +27,38 @@
...
-				</template>
-			</EmptyContent>
-
-			<File v-for="file in fileList"
-				:key="file.fileid"
-				:list="fileList"
-				v-bind="file" />
-		</template>
-	</Grid>
+		<Grid>
```
Comment (skjnldsv) [inline src/views/Tags.vue:36]: I don't see how this is relevant to this pull request?

File: src/views/Timeline.vue (hunk 5 above)
Comment (skjnldsv) [inline src/views/Timeline.vue:143]: Grid config is not used anymore

File: src/views/Timeline.vue [hunk 9]
```diff
@@ -143,47 +123,69 @@ export default {
...
-				// next time we load this script, we load the next page if the list returned
-				if (files.length === this.gridConfig.count * 5) {
-					this.page++
-				} else {
-					console.debug('We loaded the last page')
+				if (files.length === params.offset === 0) {
+					this.isEmpty = true
+				}
+
+				if (files.length !== 30) {
```
Comment (skjnldsv) [inline src/views/Timeline.vue:150]: Same here (grid config)

File: src/views/Timeline.vue [hunk 10]
```diff
@@ -143,47 +123,69 @@ export default {
...
+							list: this.$refs.virtualgrid.getCurrentItems,
+							loadMore: this.$refs.virtualgrid.loadMoreData,
+						},
+						width: 256,
+						height: 256,
+						columnSpan: 1,
+						renderComponent: FileVirtualGrid,
+					})
+					return finalArray
+				})
```
Comment (skjnldsv) [inline src/views/Timeline.vue:188]: I would love a little bit more comments here, so anyone can follow :see_no_evil: 
I prefer to have more comments than needed so at least it's clear :v:

**Review by skjnldsv — CHANGES_REQUESTED** [review]
- [x] Header Navigation is  missing when there is pictures (present when there is none)
- [ ] Loading indicator is missing
- [ ] Grid sizes config is not used anymore
- See inline comments too :)

File: src/views/Timeline.vue (hunk 5 above)
Comment (Mikescops) [reply src/views/Timeline.vue:143]: yes, it is useless has the grid is not based on row number anymore, we can decide (or let the user decide) of the number of images to load by batch.

File: src/views/Timeline.vue (hunk 9 above)
Comment (Mikescops) [reply src/views/Timeline.vue:150]: Shall we make a gridconfig that is a setting?

File: src/views/Timeline.vue (hunk 10 above)
Comment (Mikescops) [reply src/views/Timeline.vue:188]: Sure ok

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:18]: We can't remove injected, it's part of the lib. Everything inside the item is edited by the lib, `injected` is set by the user of the lib, it can be anything and the lib won't touch it, in our case it's whether an image or a separator.
If you want to better understand check out the demo of the lib.

File: src/components/SeparatorVirtualGrid.vue (hunk 7 above)
Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:40]: Yes understood

**Comment by Mikescops** [comment]
> Header Navigation is missing when there is pictures (present when there is none)

@jancborchardt asked to remove it when there is pictures

> Loading indicator is missing

I'm not sure how we can re-add the loading indicator, the lib is not ready for this, I think we can open an issue on the lib and fix this later, I'm not convinced it's a blocker

> Grid sizes config is not used anymore
As answered in the small comments, gridconfig is now handled on the lib side. It is configurable through functions passed to the grid component. In my opinion we can remove it and set the batch size with a param.

File: src/views/Tags.vue (hunk 8 above)
Comment (Mikescops) [reply src/views/Tags.vue:36]: I updated the css of the navigation to fit the new grid, and it is impacting the other pages, so I removed the navigation from being inside the grid :)

File: src/views/Timeline.vue (hunk 5 above)
Comment (Mikescops) [reply src/views/Timeline.vue:143]: reverted

File: src/views/Timeline.vue (hunk 9 above)
Comment (Mikescops) [reply src/views/Timeline.vue:150]: based on gridconfig now

File: src/views/Timeline.vue (hunk 10 above)
Comment (Mikescops) [reply src/views/Timeline.vue:188]: I added comments

File: src/components/SeparatorVirtualGrid.vue (hunk 1 above)
Comment (Mikescops) [reply src/components/SeparatorVirtualGrid.vue:18]: Removed, i misunderstood the comment sorry

**Comment by Mikescops** [comment]
I have fixed all suggested issues except the loading icon when waiting for next batch.
For this one I created an issue on the lib side: https://github.com/Mikescops/vue-virtual-grid/issues/8
If it's a non blocking regression we can move forward here.

File: src/views/Tags.vue (hunk 8 above)
Comment (Mikescops) [reply src/views/Tags.vue:36]: It is due to the new styles applied to titles to fit the new grid. I had to fix them on other pages.

**Review by skjnldsv — APPROVED** [review]

**Comment by jancborchardt** [comment]
Nice! Only found little issues:
- [ ] `.grid-container` has a white background set which shows in dark mode. It doesn’t actually need any background set? (If it does then we should use variables – but I really don’t think the background needs to be there?)
- [ ] There’s a bit too much space on the top. It’s the padding-top of the heading _plus_ the 66px of padding on the grid-container. I’d just remove the 66px top on the grid container, for any form factor. The padding of the heading is enough :)
- [ ] The bottom of the photo list has very little space, I’d go for something like 256px down there to signify the list is over.
image

**Comment by Mikescops** [comment]
@jancborchardt nice catch! done

**Review by jancborchardt — APPROVED** [review]
Looks good design-wise! :) Nice!

**Comment by Mikescops** [comment]
@skjnldsv can I let you merge this?

**Comment by TKFRvisionOfficial** [comment]
Hey guys, that is maybe a long shot but have you thought about implementing a little side timeline (a litle line with points on it on the site) where every point is a month so you can easily find pictures. Like at amazon cloud?

**Comment by Mikescops** [comment]
See #426

**Comment by sanchosk** [comment]
Would it be somehow possible to separate not only by months, but also by days?
